c++/src/kj/memory.h

```diff
@@ -656,6 +705,205 @@ class SpaceFor {
+  inline operator Ptr<T>() { return Ptr<T>(this); }
+  // Pin<T> can be implicitly converted to Ptr<T> to obtain new pointers.
+
+  inline Ptr<T> asPtr() { return Ptr<T>(this); }
+  // Explicit convenience method to create new pointers.
+
+  template <typename U, typename = EnableIf<canConvert<T*, U*>()>>
+  inline operator Ptr<U>() { return Ptr<U>(this); }
+  // Pin<T> can be implicitly converted to pointers of compatible types.
+
+  template <typename U, typename = EnableIf<canConvert<T*, U*>()>>
+  inline Ptr<U> asPtr() { return Ptr<U>(this); }
+  // Explicit convenience method to create new pointers of compatible types.
+
+  void* operator new(size_t count) = delete;
+  void* operator new[](size_t count) = delete;
+  // Pin<T> can't be heap allocated, only local or data field usage is ok.
+
+private:
+  KJ_DISALLOW_COPY(Pin);
+
+  inline Pin(T&& t): t(kj::mv(t)) {}
+
+  T t;
+#ifdef KJ_ASSERT_PTR_COUNTERS
+  _::PtrCounter ptrCounter;
+#endif
+
+  inline T* get() { return &t; }
+  inline const T* get() const { return &t; }
+
+  template <typename>
+  friend class Ptr;
+};
+
+// =======================================================================================
+// Ptr<T>
+
+template <typename T>
+class Ptr {
+// Ptr<T> is a smart alternative to T&.
+// 
+// When used together with Pin<T> it keeps track of active pointers.
+// Asserts lifetime constraints when KJ_ASSERT_PTR_COUNTERS is defined.
+// Zero-overhead alternative for T& if KJ_ASSERT_PTR_COUNTERS is not defined.
+
+public:
+  inline ~Ptr() {
+    if (ptr == nullptr) {
+      // the value was moved out
+      return;
+    }
+#ifdef KJ_ASSERT_PTR_COUNTERS
+    counter->dec();
+#endif
+  }
+
+  Ptr(Ptr&& other) : ptr(other.ptr)
+#ifdef KJ_ASSERT_PTR_COUNTERS
+  , counter(other.counter) 
+#endif
+  {
+    other.ptr = nullptr;
+  }
+
+  Ptr(const Ptr& other) : ptr(other.ptr)
+  // Ptr<T> can be freely copied.
+#ifdef KJ_ASSERT_PTR_COUNTERS
+  , counter(other.counter) 
+#endif
+  {
+#ifdef KJ_ASSERT_PTR_COUNTERS
+    counter->inc();
+#endif
+   }
+
+  inline void operator=(std::nullptr_t other) { 
+    if (ptr != nullptr) {
+#ifdef KJ_ASSERT_PTR_COUNTERS
+      counter->dec();
+      counter = nullptr;
+#endif
+      ptr = nullptr;
+    }
+  }
+
+  inline T* operator->() { return get(); }
+  inline const T* operator->() const { return get(); }
+
+  inline bool operator==(const Pin<T>& other) const { return get() == other.get(); }
+  inline bool operator==(const Ptr<T>& other) const { return get() == other.get(); }
+  inline bool operator==(const T* const other) const { return get() == other; }
+
+  template <typename U, typename = EnableIf<canConvert<U*, T*>()>>
+  inline bool operator==(const Pin<U>& other) const { return get() == other.get(); }
+
+  template <typename U, typename = EnableIf<canConvert<U*, T*>()>>
+  inline bool operator==(const Ptr<U>& other) const { return get() == other.get(); }
+
+private:
+  inline Ptr(Pin<T>* pin) : ptr(pin->get())
+#ifdef KJ_ASSERT_PTR_COUNTERS
```

Comment:
Non-blocking nitpick but I kind of think it'd be easier to read if you just had two copies of each constructor definition, one with counters enabled and one without. The partial-line ifdefs are pretty rough.